src/GovUk.Education.ExploreEducationStatistics.Content.Services/ReleaseFileService.cs

```diff
@@ -15,62 +15,100 @@
+using GovUk.Education.ExploreEducationStatistics.Content.ViewModels;
 using Microsoft.AspNetCore.Mvc;
 using Microsoft.EntityFrameworkCore;
 using static GovUk.Education.ExploreEducationStatistics.Common.BlobContainers;
 using File = System.IO.File;
 
 namespace GovUk.Education.ExploreEducationStatistics.Content.Services
 {
-    public class ReleaseFileService : IReleaseFileService
+    public class ReleaseFileService(
+        ContentDbContext contentDbContext,
+        IPersistenceHelper<ContentDbContext> persistenceHelper,
+        IPublicBlobStorageService publicBlobStorageService,
+        IDataGuidanceFileWriter dataGuidanceFileWriter,
+        IUserService userService)
+        : IReleaseFileService
     {
         /// How long the all files zip should be
         /// cached in blob storage, in seconds.
         private const int AllFilesZipTtl = 60 * 60;
 
-        private static readonly FileType[] AllowedZipFileTypes =
-        {
+        private static readonly FileType[] AllowedFileTypes =
+        [
             FileType.Ancillary,
-            FileType.Data,
-        };
-
-        private readonly ContentDbContext _contentDbContext;
-        private readonly IPersistenceHelper<ContentDbContext> _persistenceHelper;
-        private readonly IPublicBlobStorageService _publicBlobStorageService;
-        private readonly IDataGuidanceFileWriter _dataGuidanceFileWriter;
-        private readonly IUserService _userService;
-
-        public ReleaseFileService(
-            ContentDbContext contentDbContext,
-            IPersistenceHelper<ContentDbContext> persistenceHelper,
-            IPublicBlobStorageService publicBlobStorageService,
-            IDataGuidanceFileWriter dataGuidanceFileWriter,
-            IUserService userService)
+            FileType.Data
+        ];
+
+        public async Task<Either<ActionResult, IList<ReleaseFileViewModel>>> ListReleaseFiles(
+            ReleaseFileListRequest request,
+            CancellationToken cancellationToken = default)
+        {
+            var releaseFiles = await contentDbContext.ReleaseFiles
+                .Include(rf => rf.ReleaseVersion)
+                .ThenInclude(rv => rv.Publication)
+                .Include(rf => rf.File)
+                .Where(rf => AllowedFileTypes.Contains(rf.File.Type))
+                .Where(rf => request.Ids.Contains(rf.Id))
+                .Where(rv => rv.Published != null)
+                .ToListAsync(cancellationToken);
+
+            var releaseVersions = releaseFiles.Select(rf => rf.ReleaseVersion)
+                .ToHashSet();
+
+            var allowedReleaseVersions = new HashSet<ReleaseVersion>();
+
+            foreach (var releaseVersion in releaseVersions)
+            {
+                if (await userService.CheckCanViewReleaseVersion(releaseVersion).IsRight())
+                {
+                    allowedReleaseVersions.Add(releaseVersion);
+                }
+            }
```

Comment:
Did you consider returning a forbidden error if any or all of the release versions cannot be viewed?

Reply:
Nope I didn't as I've opted for a filtering approach. It'd be a little confusing for the endpoint to return a forbidden response as it's not particularly informative and won't help you pinpoint where the problem is.

Reply:
Cool, I think it's fine as it is then. Thanks for going through the reasoning.